WalletWasabi/WabiSabi/Client/ArenaClient.cs

```diff
@@ -64,6 +65,34 @@ public ArenaClient(WabiSabiClient amountCredentialClient, WabiSabiClient weightC
 			return inputRegistrationResponse.AliceId;
 		}
 
+		public async Task RemoveInputAsync(Guid roundId, Guid aliceId)
+		{
+			await RequestHandler.RemoveInputAsync(new InputsRemovalRequest(roundId, aliceId)).ConfigureAwait(false);
+		}
+
+		public async Task RegisterOutputAsync(Guid roundId, TxOut output, IEnumerable<Credential> amountCredentialsToPresent, IEnumerable<Credential> weightCredentialsToPresent)
+		{
+			var presentedAmount = amountCredentialsToPresent.Take(2).Sum(x => (long)x.Amount.ToUlong());
+			var (realAmountCredentialRequest, realAmountCredentialResponseValidation) = AmountCredentialClient.CreateRequest(
+				new[] { presentedAmount - output.Value.Satoshi },
```

Comment:
the fees are subtracted from the credential amount, so in order to register an output whose value is equal to `output.Value` the delta needs to be `(long)(-1 * (output.Value + (FeeRate/WITNESS_SCALE_FACTOR) * output.GetWeight()))`.

Speaking of which, `FeeRate` is defined in terms of decimal (how many places?) per KB, and IIRC decimals use towards-0 rounding semantics when converting to integers, which means we have to round consistently.

If `FeeRate` is an integer sat/vbyte then outputs are guaranteed to have integer fees, but inputs would need a feerate that is a multiple of 4 in sats/vbyte to work the same.

Reply:
also needs a `Guard` to ensure that `presentedAmount` is greater than the effective cost of `output`

Reply:
Calculating the fee here is something i don't want to do, at this level we should not know anything about fees. The problem is that having a `TxOut` as a parameter creates the illusion that that's the output we want to have in the coinjoin when my intend was to use it simply as a convenient container for the tuple (script, value). I will change it to eliminate this confusion.

Reply:
> also needs a Guard to ensure that presentedAmount is greater than or equal to the effective cost of output

`WabiSabiClient` already checks that.